File: .cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
using namespace std;

int main() {
    int N, M;
     cin >> N >> M;

    vector<vector<int>> routes(M);
     unordered_map<int, vector<int>> stop_to_routes;

    for (int i = 0; i < M; ++i) {
      int K;
       cin >> K;
        routes[i].resize(K);
          for (int j = 0; j < K; ++j) {
            cin >> routes[i][j];
             stop_to_routes[routes[i][j]].push_back(i);
        }
    }

    int A, B;
     cin >> A >> B;

       if (A == B) {
        cout << 0 << endl;
         return 0;
    }

    queue<pair<int, int>> q;
     unordered_set<int> visited_stops;
     unordered_set<int> visited_routes;

    // Запускаем BFS с остановки A
    for (int route : stop_to_routes[A]) {
      q.push({route, 0});
       visited_routes.insert(route);
    }
      visited_stops.insert(A);

    while (!q.empty()) {
      auto [current_route, transfers] = q.front();
       q.pop();

        for (int stop : routes[current_route]) {
          if (stop == B) {
            cout << transfers << endl;
             return 0;
            }
            if (visited_stops.find(stop) == visited_stops.end()) {
              visited_stops.insert(stop);
                for (int next_route : stop_to_routes[stop]) {
                    if (visited_routes.find(next_route) == visited_routes.end()) {
                     q.push({next_route, transfers + 1});
                     visited_routes.insert(next_route);
                    }
                }
            }
        }
    }

    cout << "Call a taxi!" << endl;
    return 0;
}

File: .js
process.stdin.resume();
process.stdin.setEncoding('utf-8');

let input = '';
process.stdin.on('data', function(chunk) {
 input += chunk;
});

process.stdin.on('end', function() {
  const lines = input.trim().split('\n');
  const [N, M] = lines[0].split(' ').map(Number);

    const routes = [];
    const stopToRoutes = new Map();

    for (let i = 1; i <= M; ++i) {
     const routeStops = lines[i].split(' ').slice(1).map(Number);
        routes.push(routeStops);
          for (const stop of routeStops) {
            if (!stopToRoutes.has(stop)) {
              stopToRoutes.set(stop, []);
            }
               stopToRoutes.get(stop).push(i - 1);
        }
    }

    const [A, B] = lines[M + 1].split(' ').map(Number);

      if (A === B) {
       console.log(0);
        return;
    }

    const queue = [];
     const visitedStops = new Set();
     const visitedRoutes = new Set();

    for (const route of (stopToRoutes.get(A) || [])) {
      queue.push([route, 0]);
       visitedRoutes.add(route);
         }
        visitedStops.add(A);

    while (queue.length > 0) {
     const [currentRoute, transfers] = queue.shift();

         for (const stop of routes[currentRoute]) {
            if (stop === B) {
                console.log(transfers);
                return;
               }
              if (!visitedStops.has(stop)) {
                visitedStops.add(stop);
                for (const nextRoute of (stopToRoutes.get(stop) || [])) {
                    if (!visitedRoutes.has(nextRoute)) {
                        queue.push([nextRoute, transfers + 1]);
                        visitedRoutes.add(nextRoute);
                    }
                }
            }
        }
    }

    console.log("Call a taxi!");
});
